### PR TITLE
Add support for named child routes

### LINKED ISSUE #55: Support named child routes
React-router provides a way to provide multiple components for a single route. It would be nice if found would match this functionality.

### PR DESCRIPTION
**Pull request by taion**
Closes #55 

How does this look?

Docs forthcoming, but the tests mostly show how the code works (though the nav definitions are a bit flaky).

Found Relay will mostly work out-of-the-box, except the param accumulation can be made more correct (it should be group-aware, since subsequent elements in the route array are now no longer always children).

### CODE REVIEW

File: src/utils/resolveRenderArgs.js [hunk 1]
```diff
@@ -1,5 +1,27 @@
 import HttpError from '../HttpError';
 
+function foldElements(elementsRaw, routeIndices) {
```
Comment (taion) [inline src/utils/resolveRenderArgs.js]: Is this the right place for this logic? I didn't want to change the signature for `<ElementsRenderer>`.

**Comment by hannesj** [comment]
How does this work with unbalanced(?) groups

I would like to have the following structure, 

```
<Route
  path="/"
  Component={Layout}
  groups={{
    title: [
      {
        Component: Title,
      },
    ],
    content: [
      {
        Component: IndexPage
        children: [
          {
            path: 'nearby',
            getComponent: () =>
              import('./component/NearbyPanel')
                .then(getDefault)
                .catch(errorLoading),
          },
          {
            path: 'favourites',
            getComponent: () =>
              import('./component/FavouritesPanel')
                .then(getDefault)
                .catch(errorLoading),
          },
        ],
      },
    ],
  }}
/>
```

It works fine when on route `/`, but when going to either `/nearby` or `/favourites`, it doesn't match as the title doesn't match anymore, and so the content group is not tested for matching? I need to have the `path="/"` on the topmost level as it has some siblings with different layouts, which can't use the same top level layout component.

**Comment by taion** [comment]
@hannesj 

It should be:

```js
{
  Component: Title,
  children: [{ path: '*' }],
},
```

It's a tad clunky but it works. I think for whatever reason we decided not to have `*` match an empty path: https://github.com/4Catalyzer/found/pull/100#discussion_r122830756.

**Comment by taion** [comment]
I guess I should have used that pattern in the tests \:

**Comment by hannesj** [comment]
Thanks, that seems to do the trick!

**Comment by taion** [comment]
Docs and better JSX support added. This is ready to go now, if anybody wants to take a look.

Barring blockers, I'll merge and cut a release tomorrow.

**Comment by taion** [comment]
@hannesj 

Turns out you can actually just do:

```js
route = {
  path: '(.*)?',
  Component: Title,
};
```

Which is (I think) a little bit cleaner.

**Comment by taion** [comment]
Wait a sec. There's no reason that we can't take an object for `children` for named child routes, instead of using the `groups` prop. The problems with using objects as children in React don't arise unless you try to e.g. map the object child, or to render it, but that doesn't arise when using these only for configuration.

@jquense @itajaja Thoughts?

**Comment by taion** [comment]
I've tweaked the API somewhat to use an object for children instead of the `groups` prop. It's a little bit unidiomatic for JSX, but we're doing a bunch of weird things with JSX config as-is, and ends up looking a bit better to my eye.

**Pull request merged**